### PR TITLE
add Pac-Man (the game) app

### PR DESCRIPTION
**Pull request by theofficialgman**
closes https://github.com/Botspot/pi-apps/issues/1356 potentially closes https://github.com/Botspot/pi-apps/issues/1327

blocked by bug at @techcoder20 https://github.com/techcoder20/RPIDoom3Installer/tree/main/SDLAddOns

### CODE REVIEW

**Comment by cycool29** [comment]
https://github.com/Botspot/pi-apps/issues/1327

**Comment by theofficialgman** [comment]
@cycool29 ahah didn't see that
well this isn't a terminal based pacman
its a sdl2 based pacman with gamecontroller, mouse and keyboard support

**Comment by cycool29** [comment]
There are already prebuilt pacman binaries in the APT repo.

I don't play PacMan, and I don't know which is better between the SDL version and the APT repo version.

**Comment by theofficialgman** [comment]
> There are already prebuilt pacman binaries in the APT repo.
> 
> I don't play PacMan, and I don't know which is better between the SDL version and the APT repo version.

the apt version is a terminal based game:
image

the pacman I have here is from this github: https://github.com/ebuc99/pacman
its fully featured and should play (and look similar) to the original pacman game with support for game controllers, etc
image

**Comment by cycool29** [comment]
> the apt version is a terminal based game:


Ahhh I just realized the game is in a terminal... lol.


Just installed the SDL one, looks better than the APT repo one, especially the sound effect.

File: apps/Pac-Man/install [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+cd /tmp || error "Could not move to tmp folder"
+
+install_packages libsdl2-dev libsdl2-ttf-dev libsdl2-image-dev libsdl2-mixer-dev || error "Could not install dependencies"
```
Comment (cycool29) [inline apps/Pac-Man/install:5]:
```suggestion
install_packages libsdl2-dev libsdl2-ttf-dev libsdl2-image-dev libsdl2-mixer-dev || exit 1
```

File: apps/Pac-Man/install [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+rm -rf /tmp/pacman
+
+git_clone https://github.com/ebuc99/pacman || error "Could not clone source"
```
Comment (cycool29) [inline apps/Pac-Man/install:8]:
```suggestion
git_clone https://github.com/ebuc99/pacman || exit 1
```

File: apps/Pac-Man/uninstall [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+sudo rm -f "/usr/local/bin/pacman"
+sudo rm -rf "/usr/local/share/pacman"
+sudo rm -f "/usr/local/share/applications/pacman.desktop"
+sudo rm -rf "/usr/local/share/doc/pacman"
+
+sudo update-desktop-database -q
+
+purge_packages || error "Dependencies failed to uninstall"
```
Comment (cycool29) [inline apps/Pac-Man/uninstall:10]:
```suggestion
purge_packages || exit 1
```

Comment (theofficialgman) [reply apps/Pac-Man/uninstall:10]: is there a reason why you want me to just exit and error without telling the user what went wrong?
(this goes for all your suggestions)

Comment (cycool29) [reply apps/Pac-Man/uninstall:10]: This is suggested by Botspot: https://github.com/Botspot/pi-apps/pull/1139#issuecomment-969705180

**Comment by cycool29** [comment]
I think we should use the icon from the menu shortcut.

24x24:
icon-24

64x64:
icon-64

**Comment by theofficialgman** [comment]
> I think we should use the icon from the menu shortcut.

I am using the one from the menu shortcut https://github.com/ebuc99/pacman/blob/master/data/gfx/pacman_desktop.png ....
where did you get that one from, is that from the other pacman app?

**Comment by cycool29** [comment]
> > I think we should use the icon from the menu shortcut.
> 
> I am using the one from the menu shortcut https://github.com/ebuc99/pacman/blob/master/data/gfx/pacman_desktop.png .... where did you get that one from, is that from the other pacman app?

image

**Comment by theofficialgman** [comment]
idk how you are getting that one... thats the old icon from the github repo. It was changed in 2019
maybe you have a theme that is overwriting it somehow

regardless thats not the one that gets installed
Screenshot from 2022-01-06 23-55-46

**Comment by cycool29** [comment]
I use the zip file from its release page. Maybe it is outdated.

When I try to install from this PR, I got this error:

```
OS: Raspbian GNU/Linux 11 (bullseye)
OS architecture: 32-bit
Pi-Apps local commit ID: c4d4fd3319342769b971381f22b467e55f5ded70
Github latest commit ID: 779aae10235ca41317ce874a13e407ccc7aeb57b	
Kernel: aarch64 5.10.63-v8+
Device model: Raspberry Pi 4 Model B Rev 1.4
Ram size: 7.81 GB
Raspberry Pi OS image version: 2021-10-30
Language: en_US.UTF-8

BEGINNING OF LOG FILE:
-----------------------

Will install these packages: libsdl2-dev libsdl2-ttf-dev libsdl2-image-dev libsdl2-mixer-dev 
Running sudo apt update...
Hit:1 https://itai-nelken.github.io/weekly-box86-debs/debian  InRelease
Hit:2 http://archive.raspberrypi.org/debian bullseye InRelease
Hit:3 http://raspbian.raspberrypi.org/raspbian bullseye InRelease
All packages are up to date.
apt update complete.
Creating an empty apt-package to install the necessary apt packages...
It will be named: pi-apps-84245b83
Depends: libsdl2-dev, libsdl2-image-dev, libsdl2-mixer-dev, libsdl2-ttf-dev
Installing the pi-apps-84245b83 package...
The following additional packages will be installed:
  libglu1-mesa-dev libibus-1.0-dev libopusfile0 libsdl2-dev libsdl2-image-2.0-0
  libsdl2-image-dev libsdl2-mixer-2.0-0 libsdl2-mixer-dev libsdl2-ttf-2.0-0
  libsdl2-ttf-dev libsndio-dev libxv-dev libxxf86vm-dev
  x11proto-xf86vidmode-dev
The following NEW packages will be installed:
  libglu1-mesa-dev libibus-1.0-dev libopusfile0 libsdl2-dev libsdl2-image-2.0-0
  libsdl2-image-dev libsdl2-mixer-2.0-0 libsdl2-mixer-dev libsdl2-ttf-2.0-0
  libsdl2-ttf-dev libsndio-dev libxv-dev libxxf86vm-dev pi-apps-84245b83
  x11proto-xf86vidmode-dev
0 upgraded, 15 newly installed, 0 to remove and 0 not upgraded.
(Reading database ... 
(Reading database ... 5%
(Reading database ... 10%
(Reading database ... 15%
(Reading database ... 20%
(Reading database ... 25%
(Reading database ... 30%
(Reading database ... 35%
(Reading database ... 40%
(Reading database ... 45%
(Reading database ... 50%
(Reading database ... 55%
(Reading database ... 60%
(Reading database ... 65%
(Reading database ... 70%
(Reading database ... 75%
(Reading database ... 80%
(Reading database ... 85%
(Reading database ... 90%
(Reading database ... 95%
(Reading database ... 100%
(Reading database ... 269065 files and directories currently installed.)

dpkg: error processing archive /tmp/apt-dpkg-install-pTJ4YL/09-libsdl2-image-dev_2.0.5+dfsg1-2_armhf.deb (--unpack):

 trying to overwrite '/usr/include/SDL2/SDL_image.h', which is also in package sdl2-image 2.0.5-1

dpkg: error processing archive /tmp/apt-dpkg-install-pTJ4YL/11-libsdl2-mixer-dev_2.0.4+dfsg1-3_armhf.deb (--unpack):

 trying to overwrite '/usr/include/SDL2/SDL_mixer.h', which is also in package sdl2-mixer 2.0.4-1

dpkg: error processing archive /tmp/apt-dpkg-install-pTJ4YL/13-libsdl2-ttf-dev_2.0.15+dfsg1-1_armhf.deb (--unpack):

 trying to overwrite '/usr/include/SDL2/SDL_ttf.h', which is also in package sdl2-ttf 2.0.15-1

Errors were encountered while processing:

 /tmp/apt-dpkg-install-pTJ4YL/09-libsdl2-image-dev_2.0.5+dfsg1-2_armhf.deb

 /tmp/apt-dpkg-install-pTJ4YL/11-libsdl2-mixer-dev_2.0.4+dfsg1-3_armhf.deb

 /tmp/apt-dpkg-install-pTJ4YL/13-libsdl2-ttf-dev_2.0.15+dfsg1-1_armhf.deb

E: Sub-process /usr/bin/dpkg returned an error code (1)
Apt finished.
Failed to install the packages!
APT reported these errors:
E: Sub-process /usr/bin/dpkg returned an error code (1)
The following additional packages will be installed:
  libglu1-mesa-dev libibus-1.0-dev libopusfile0 libsdl2-dev libsdl2-image-2.0-0
  libsdl2-image-dev libsdl2-mixer-2.0-0 libsdl2-mixer-dev libsdl2-ttf-2.0-0
  libsdl2-ttf-dev libsndio-dev libxv-dev libxxf86vm-dev
  x11proto-xf86vidmode-dev
The following NEW packages will be installed:
  libglu1-mesa-dev libibus-1.0-dev libopusfile0 libsdl2-dev libsdl2-image-2.0-0
  libsdl2-image-dev libsdl2-mixer-2.0-0 libsdl2-mixer-dev libsdl2-ttf-2.0-0
  libsdl2-ttf-dev libsndio-dev libxv-dev libxxf86vm-dev pi-apps-84245b83
  x11proto-xf86vidmode-dev
0 upgraded, 15 newly installed, 0 to remove and 0 not upgraded.
(Reading database ... 
(Reading database ... 5%
(Reading database ... 10%
(Reading database ... 15%
(Reading database ... 20%
(Reading database ... 25%
(Reading database ... 30%
(Reading database ... 35%
(Reading database ... 40%
(Reading database ... 45%
(Reading database ... 50%
(Reading database ... 55%
(Reading database ... 60%
(Reading database ... 65%
(Reading database ... 70%
(Reading database ... 75%
(Reading database ... 80%
(Reading database ... 85%
(Reading database ... 90%
(Reading database ... 95%
(Reading database ... 100%
(Reading database ... 269065 files and directories currently installed.)

dpkg: error processing archive /tmp/apt-dpkg-install-pTJ4YL/09-libsdl2-image-dev_2.0.5+dfsg1-2_armhf.deb (--unpack):

 trying to overwrite '/usr/include/SDL2/SDL_image.h', which is also in package sdl2-image 2.0.5-1

dpkg: error processing archive /tmp/apt-dpkg-install-pTJ4YL/11-libsdl2-mixer-dev_2.0.4+dfsg1-3_armhf.deb (--unpack):

 trying to overwrite '/usr/include/SDL2/SDL_mixer.h', which is also in package sdl2-mixer 2.0.4-1

dpkg: error processing archive /tmp/apt-dpkg-install-pTJ4YL/13-libsdl2-ttf-dev_2.0.15+dfsg1-1_armhf.deb (--unpack):

 trying to overwrite '/usr/include/SDL2/SDL_ttf.h', which is also in package sdl2-ttf 2.0.15-1

Errors were encountered while processing:

 /tmp/apt-dpkg-install-pTJ4YL/09-libsdl2-image-dev_2.0.5+dfsg1-2_armhf.deb

 /tmp/apt-dpkg-install-pTJ4YL/11-libsdl2-mixer-dev_2.0.4+dfsg1-3_armhf.deb

 /tmp/apt-dpkg-install-pTJ4YL/13-libsdl2-ttf-dev_2.0.15+dfsg1-1_armhf.deb

E: Sub-process /usr/bin/dpkg returned an error code (1)

Failed to install Pac-Man!
◢◣Need help? Copy the ENTIRE terminal output or take a screenshot.
Please ask on Github: https://github.com/Botspot/pi-apps/issues/new/choose
Or on Discord: https://discord.gg/RXSTvaUvuu
```

**Comment by theofficialgman** [comment]
```
dpkg: error processing archive /tmp/apt-dpkg-install-pTJ4YL/09-libsdl2-image-dev_2.0.5+dfsg1-2_armhf.deb (--unpack):

 trying to overwrite '/usr/include/SDL2/SDL_image.h', which is also in package sdl2-image 2.0.5-1

dpkg: error processing archive /tmp/apt-dpkg-install-pTJ4YL/11-libsdl2-mixer-dev_2.0.4+dfsg1-3_armhf.deb (--unpack):

 trying to overwrite '/usr/include/SDL2/SDL_mixer.h', which is also in package sdl2-mixer 2.0.4-1

dpkg: error processing archive /tmp/apt-dpkg-install-pTJ4YL/13-libsdl2-ttf-dev_2.0.15+dfsg1-1_armhf.deb (--unpack):

 trying to overwrite '/usr/include/SDL2/SDL_ttf.h', which is also in package sdl2-ttf 2.0.15-1

Errors were encountered while processing:

 /tmp/apt-dpkg-install-pTJ4YL/09-libsdl2-image-dev_2.0.5+dfsg1-2_armhf.deb

 /tmp/apt-dpkg-install-pTJ4YL/11-libsdl2-mixer-dev_2.0.4+dfsg1-3_armhf.deb

 /tmp/apt-dpkg-install-pTJ4YL/13-libsdl2-ttf-dev_2.0.15+dfsg1-1_armhf.deb
 ```
 well... I can't help if raspbian bullseye has broken sdl2 dev packages.... go complain to the raspbian maintainers to fix those, they broken them.
 works perfectly find on debian bullseye
 

>  I use the zip file from its release page. Maybe it is outdated.

yes the zip file from the release page is out of date

**Comment by theofficialgman** [comment]
@cycool29 just to make sure... could you report the results from this here for me?
```
apt list --installed | grep -E "sdl2-image|sdl2-mixer|sdl2-ttf"
```

**Comment by cycool29** [comment]
> @cycool29 just to make sure... could you report the results from this here for me?
> 
> ```
> apt list --installed | grep -E "sdl2-image|sdl2-mixer|sdl2-ttf"
> ```

Sure.

```
sdl2-image/now 2.0.5-1 armhf [installed]
sdl2-mixer/now 2.0.4-1 armhf [installed]
sdl2-ttf/now 2.0.15-1 armhf [installed]
```

**Comment by theofficialgman** [comment]
ok so its a you problem
those packages do not exist on debian/raspbian or ubuntu
those are manually installed packages from somewhere....
any clue where you got them? because they are breaking the apt repo packages and need to be removed

**Comment by theofficialgman** [comment]
run an apt cache policy on each of them
```
apt-cache policy sdl2-image

apt-cache policy sdl2-mixer

apt-cache policy sdl2-ttf
```

**Comment by cycool29** [comment]
> any clue where you got them? because they are breaking the apt repo packages and need to be removed

Not sure. Removing them doesn't break my system, so I removed them. 

And now the installation of Pacman goes well.

**Comment by theofficialgman** [comment]
> > any clue where you got them? because they are breaking the apt repo packages and need to be removed
> 
> Not sure. Removing them doesn't break my system, so I removed them.

noo... why did you remove them, I wanted to know how they got there, lol
oh well, it someone else has an issue with these same packages we will see in the future

**Comment by cycool29** [comment]
Now I know where I got those packages - Doom 3.

https://github.com/techcoder20/RPIDoom3Installer/tree/main/SDLAddOns

**Comment by theofficialgman** [comment]
> Now I know where I got those packages - Doom 3.
> 
> https://github.com/techcoder20/RPIDoom3Installer/tree/main/SDLAddOns

@cycool29 thank you for posting that, that will need to be changed as its VERY bad and dangerous as you can see

**Comment by ryanfortner** [comment]
If you want to create deb packages, I've added the necessary files to do so in my fork. Here are the instructions:

```
sudo apt install devscripts -y
git clone https://github.com/ryanfortner/pacman.git
cd pacman
sudo mk-build-deps -i debian/control
sudo rm pacman-build-deps*
dpkg-buildpackage --no-sign
# package will be in the parent directory
```

**Comment by theofficialgman** [comment]
> If you want to create deb packages, I've added the necessary files to do so in my fork. Here are the instructions:

I do not prefer to make deb packages for the reason I already stated above:

>  (also I don't prefer debs because that locks in what systems they can run on.... pretty much anything you compile on buster makes it incompatibly with bionic due to newer glibc)

there isn't any reason this can't build on bionic/stretch/xenial, etc. building a deb on buster is just going to lock in glibc to 2.28+. If you want to make debs, try to build on the OLDEST OS you can (stretch or xenial would be a start) and verify that the deb still works on anything newer

**Comment by ryanfortner** [comment]
I built debs, they are located here: https://apt.raspbian-addons.org/debian/pool/main/p/pacman/

**Comment by theofficialgman** [comment]
> I built debs, they are located here: https://apt.raspbian-addons.org/debian/pool/main/p/pacman/
@ryanfortner 
See my comment about yours. What did you build using?

**Comment by theofficialgman** [comment]
> I built debs, they are located here: https://apt.raspbian-addons.org/debian/pool/main/p/pacman/

@ryanfortner 
See my comment about yours. What did you build using?

**Comment by ryanfortner** [comment]
> @ryanfortner See my comment above yours. What did you build using?

Debian Buster. I don't currently have an installation of stretch/xenial/bionic or any other older distro unfortunately

**Comment by derryktutt** [comment]
> > @ryanfortner See my comment above yours. What did you build using?
> 
> Debian Buster. I don't currently have an installation of stretch/xenial/bionic or any other older distro unfortunately

I do, somewhere. I cant find it, though.

**Comment by Botspot** [comment]
This window had a very strange behavior when trying to close it. To go back to the menu after finishing a game, I have to press the window-manager Close button. Looks like this version of PacMan treats it as a Back button.

Is there a keyboard shortcut for going back? Or is this the intended way to do it? In any case, I think it would be wise to mention this in the description.

**Comment by theofficialgman** [comment]
> This installed fine for me. This window had a very strange behavior when trying to close it. To go back to the menu after finishing a game, I have to press the window-manager Close button. Looks like this version of PacMan treats it as a Back button.

this is intentional, its how they coded the game. the window manager close button, the minus button on a controller, and the ESC key on a keyboard function as the back/quit button. when on the main menu, these buttons quit the game.

> Is there a keyboard shortcut for going back? Or is this the intended way to do it? In any case, I think it would be wise to mention this in the description.

ESC key (see above)

**Pull request merged**